docs/source/user-guide/arrow-introduction.md

```diff
@@ -0,0 +1,284 @@
+
+fn make_batch() -> arrow_schema::Result<RecordBatch> {
+    let ids = Int32Array::from(vec![1, 2, 3]);
+    let names = StringArray::from(vec![Some("alice"), None, Some("carol")]);
+
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("id", DataType::Int32, false),
+        Field::new("name", DataType::Utf8, true),
+    ]));
+
+    let cols: Vec<ArrayRef> = vec![Arc::new(ids), Arc::new(names)];
+    RecordBatch::try_new(schema, cols)
+}
+```
+
+## Query an in-memory batch with DataFusion
+
+Once you have a [`RecordBatch`], you can query it with DataFusion using a [`MemTable`]. This is useful for testing, processing data from external systems, or combining in-memory data with other sources. The example below creates a batch, wraps it in a [`MemTable`], registers it as a named table, and queries it using SQL—demonstrating how Arrow serves as the bridge between your data and DataFusion's query engine.
+
+```rust
+use std::sync::Arc;
+use arrow_array::{Int32Array, StringArray, RecordBatch};
+use arrow_schema::{DataType, Field, Schema};
+use datafusion::datasource::MemTable;
+use datafusion::prelude::*;
+
+#[tokio::main]
+async fn main() -> datafusion::error::Result<()> {
+    let ctx = SessionContext::new();
+
+    // build a batch
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("id", DataType::Int32, false),
+        Field::new("name", DataType::Utf8, true),
+    ]));
+    let batch = RecordBatch::try_new(
+        schema.clone(),
+        vec![
+            Arc::new(Int32Array::from(vec![1, 2, 3])) as _,
+            Arc::new(StringArray::from(vec![Some("foo"), Some("bar"), None])) as _,
+        ],
+    )?;
+
+    // expose it as a table
+    let table = MemTable::try_new(schema, vec![vec![batch]])?;
+    ctx.register_table("people", Arc::new(table))?;
+
+    // query it
+    let df = ctx.sql("SELECT id, upper(name) AS name FROM people WHERE id >= 2").await?;
+    df.show().await?;
+    Ok(())
+}
+```
+
+## Common Pitfalls
+
+When working with Arrow and RecordBatches, watch out for these common issues:
+
+- **Schema consistency**: All batches in a stream must share the exact same [`Schema`]. For example, you can't have one batch where a column is [`Int32`] and the next where it's [`Int64`], even if the values would fit
+- **Immutability**: Arrays are immutable—to "modify" data, you must build new arrays or new RecordBatches. For instance, to change a value in an array, you'd create a new array with the updated value
+- **Buffer management**: Variable-length types (UTF-8, binary, lists) use offsets + values arrays internally. Avoid manual buffer slicing unless you understand Arrow's internal invariants—use Arrow's built-in compute functions instead
+- **Type mismatches**: Mixed input types across files may require explicit casts. For example, a string column `"123"` from a CSV file won't automatically join with an integer column `123` from a Parquet file—you'll need to cast one to match the other
```

Comment:
Maybe link to the cast kernel: https://docs.rs/arrow/latest/arrow/compute/fn.cast.html